src/runtime_src/core/edge/skd/xrt_skd.cpp

```diff
@@ -35,313 +40,370 @@ namespace xrt {
-      return -1;
+      msg = std::string("Cannot create soft kernel.");
+      xrt_core::message::send(severity_level::error, "SKD", msg);
+      return ret;
     }
 
-    syslog(LOG_INFO, "%s_%d start running, cmd_boh = %d\n", sk_name, cu_idx, cmd_boh);
+    msg = m_sk_name + std::to_string(m_cu_idx) + std::string(" start running, cmd_boh = ") + std::to_string(m_cmd_boh);
+    xrt_core::message::send(severity_level::info, "SKD", msg);
 
-    args_from_host = (unsigned *)xclMapBO(devHdl, cmd_boh, true);;
-    if (args_from_host == MAP_FAILED) {
-      syslog(LOG_ERR, "Failed to map soft kernel args for %s_%d", sk_name, cu_idx);
-      dlclose(sk_handle);
-      return -1;
+    m_args_from_host = reinterpret_cast<unsigned *>(xclMapBO(m_devhdl, m_cmd_boh, true));
+    if (m_args_from_host == MAP_FAILED) {
+      msg = std::string("Failed to map soft kernel args for ") + m_sk_name + std::to_string(m_cu_idx);
+      xrt_core::message::send(severity_level::error, "SKD", msg);
+      dlclose(m_sk_handle);
+      return -EINVAL;
     }
 
     // Prep FFI type for all kernel arguments
-    for(int i=0;i<num_args;i++) {
-      ffi_args[i] = convert_to_ffitype(args[i]);
+    for(const auto &i : m_kernel_args) {
+      m_ffi_args.emplace_back(convert_to_ffitype(i));
     }
 
-    if(ffi_prep_cif(&cif,FFI_DEFAULT_ABI, num_args, &ffi_type_uint32,ffi_args) != FFI_OK) {
-      syslog(LOG_ERR, "Cannot prep FFI arguments!");
-      return -1;
+    // Expect PS kernels to return POSIX return code 
+    if(ffi_prep_cif(&m_cif,FFI_DEFAULT_ABI, m_kernel_args.size(), &ffi_type_uint32, m_ffi_args.data()) != FFI_OK) {
+      const std::string errMsg = "Cannot prep FFI arguments!";
+      xrt_core::message::send(severity_level::error, "SKD", errMsg);
+      return -EINVAL;
     }
 
-    syslog(LOG_INFO,"Finish soft kernel %s init\n",sk_name);
+    msg = std::string("Finish soft kernel %s init\n") + m_sk_name;
+    xrt_core::message::send(severity_level::debug, "SKD", msg);
     return 0;
   }
 
   XCL_DRIVER_DLLESPEC
   void
   skd::run() {
-    int32_t kernel_return = 0;
+    ffi_arg kernel_return = 0;
     int ret = 0;
-    void* ffi_arg_values[num_args];
+    std::vector<void*> ffi_arg_values(m_kernel_args.size());
     // Buffer Objects
-    int boHandles[num_args];
-    void* bos[num_args];
-    uint64_t boSize[num_args];
+    std::vector<size_t> bo_offsets(m_kernel_args.size());
+    std::vector<void*> bos(m_kernel_args.size());
+    std::vector<size_t> bo_size(m_kernel_args.size());
+    std::vector<int> bo_handles(m_kernel_args.size());
     std::vector<int> bo_list;
-
-    while (1) {
-      ret = waitNextCmd();
-
-      if (ret) {
-	/* We are told to exit the soft kernel loop */
-	syslog(LOG_INFO, "Exit soft kernel %s\n", sk_name);
-	break;
+    Clock::time_point start;
+    Clock::time_point end;
+    Clock::time_point cmd_start;
+    Clock::time_point cmd_end;
+
+    while (true) {
+      ret = wait_next_cmd();
```

Comment:
Hmm...what is the variable `ret `used for in this while look.

Reply:
It's used to check the exit condition